enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,821 @@
+update of the `.status` of the CR.
+* The controller will also store the wildcard DNS name which matches the `.spec.name` field of a `DNSNameResolver` CR. If the DNS lookup is
+successful for the wildcard DNS name, then the DNS name will be stored with the corresponding current IP addresses along with the TTL and the last lookup time
+for each IP address. If the DNS lookup is not successful, then it will be retried after a default TTL (30 minutes). The `DNSNameResolver` controller will
+also store the regular DNS names, matching the wildcard DNS name's `.status.resolvedNames[*].dnsName` field, with the corresponding current IP addresses along
+with the TTL and the last time to lookup for each IP address. Using the TTL and last lookup time of all the associated IP addresses of a DNS name, a next lookup
+time for the DNS name is also stored. Based on the next lookup time, the controller will follow the same method as that of the regular DNS names to get the
+latest IP addresses and TTL information.
+* The controller will also check if any IP address is needed to be removed from the `.status` field. The validity of any IP address expires after the
+corresponding next lookup time (TTL + last lookup time). However, the IP address may still be in use. Thus, a grace period of 1 second is provided for
+each IP address and on expiration of the grace period the IP address will be removed. If any IP address satisfies this condition then it is removed from
+the `.status` of the `DNSNameResolver` CR and the object is updated.
+* If the validity (TTL + last lookup time) of any of the IP addresses have expired, but the grace period of 1 second is still not over, then the same event
+will be reconciled after the grace period has expired. This is done so that during the next reconcile the IP addresses can be removed from the `.status` of the
+`DNSNameResolver` CR.
+* On receiving the delete event, the `DNSNameResolver` controller will remove the details stored regarding the `DNSNameResolver` CRs.
+* If the deleted event is for a `DNSNameResolver` corresponding to a regular DNS name, the `DNSNameResolver` controller will remove the details of a regular DNS
+name, if it doesn't match a `DNSNameResolver` CR corresponding to a  wildcard DNS name.
+* If the deleted event is for a `DNSNameResolver` corresponding to a wildcard DNS name, the `DNSNameResolver` controller will remove the details of a wildcard DNS
+name. It will also remove the details of the regular DNS names added to the `.status` of the CR, if each of them doesn't match a `DNSNameResolver` CR corresponding
+to a  wildcard DNS name.
+
+#### Workflow of `ocp_dnsnameresolver` CoreDNS plugin
+
+The following pseudocodes explain the workflow of the `ocp_dnsnameresolver` CoreDNS plugin:
+
+##### Watch
+```shell
+watch(DNSNameResolver)
+
+if match(event.type, "create"):
```

Comment:
This watches updates too, right?

Reply:
Here only the `spec.name` field of the DNSNameResolver CR is being used. This field stores the DNS name and it is immutable. Thus, watching updates events will not be required.